### ISSUE TITLE
Request: `cleanup` coroutine in SSHServer 

### ISSUE THREAD

**Issue by Kentzo** [ronf/asyncssh#602]
I'd like to allocate some asyncio-managed resources to be shared among all server's sessions.

**Comment by ronf** [comment]
The `connection_made()` and `connection_lost()` methods are part of the asyncio callback-based API defined in `asyncio.BaseProtocol`. Making these methods async is not really an option. However, there are other ways to accomplish what I think you're asking for.

If you are using the high-level APIs like `create_server()` or `listen()`, you can do whatever you'd want to do at the top of whatever callable or coroutine you pass in as the `process_factory` or `session_factory` that you tell it to run when a new session is started, if your goal is to share state across sessions.

If you want to share state across _connections_, then you could create a custom `SSHServer` subclass and pass that into `asyncssh.create_server()` as the `server_factory`. This can be initialized with class members holding the shared state prior to creating the SSH listener which uses it, and those calls could be coroutines if they needed to be. The resulting `SSHServer` instances can be queried by calling `get_owner()` on the `SSHServerConnection objects that are opened.

If you want to run some code exactly once every time an SSH connection's authentication finishes (as opposed to on every session within the connection) or when there's an error on a connection, there are arguments to pass in an `acceptor` and an `error_handler` in the `SSHServerConnectionOptions`. The former runs when auth completes successfully on a connection and the latter runs whenever a connection is being cleaned up, allowing you to see an exception which triggered it to close.

**Comment by Kentzo** [comment]
Ah, interesting.

May I then request to improve the documentation of the `error_handler` argument to reflect that it's closed for normally closed connections as well?

**Comment by ronf** [comment]
Sorry, I should have been more precise in my description above.

When authentication completes successfully on a new SSH connection, the `acceptor` callable/coroutine is called, and the error handler is cleared without ever being called. The `error_handler` is only called if there is a failure before authentication is successful on the connection. In this case, the `acceptor` will not be called and instead the `error_handler` is called with the exception information for what triggered the failure. For any given connection one or the other will be called, but never both.

The error handler won't be called when a connection is closed normally, unless perhaps the close happens before authentication completes, in which case I would expect it to report a ConnectionLost exception. It could also potentially report a DisconnectError exception if the client sent an SSH disconnect message prior to closing the TCP connection.

**Comment by Kentzo** [comment]
I was hoping to allocate an awaitable resource in `SSHServer.validate_password`, because it depends on the provided username and password and is shared by all sessions. But then I don't see a trivial way to clean up, other than blocking in `SSHServer.connection_lost` (tbh not sure if that would work at all).

**Comment by ronf** [comment]
The `validate_password()` method actually can support being either a callable or a coroutine, meaning that you can declare it as "async" and do "blocking" operations in it if you need to which won't prevent other tasks from making progress.

This is not true of the `connection_lost()` method -- it can only be a callable for now. However, if you have a stored copy of the connection object, you can call `conn.create_task()` to schedule a new task to run from within `connection_lost()`, and that can do blocking operations. Perhaps that'll be enough to do what you're looking to do.

**Comment by Kentzo** [comment]
I can come up with a workaround, yes. However it would be nice to have the cleanup asynchronous method somewhere within SSHServer.

**Comment by ronf** [comment]
It would be possible to allow `connection_lost()` to be an optional coroutine, similar to what was done with auth, but I'm hesitant to do that, since that class is currently modeled to look like `asyncio.Protocol`, even though it doesn't directly inherit from that, and the definition in `asyncio.Protocol` doesn't allow the method to be a coroutine. I did this for the auth methods, but those were unique to AsyncSSH.

It's really quite clean to use `create_task()` for this. Basically, instead of:

```python
    await do_some_cleanup(args)
```

you'd just do:

```python
    self._conn.create_task(do_some_cleanup(args))
```

...assuming that `self._conn` is assigned when `connection_made()` is called with the SSHConnection instance.

You could even use `asyncio.ensure_future()` to create a new task without referencing the connection object, but there's some additional exception handling which SSHConnection's `create_task()` adds that could be beneficial here.

**Comment by ronf** [comment]
I guess one possible alternative would be to add another handler argument, alongside the acceptor and error_handler. It could be called something like close_handler, running when a successfully opened connection closes. In other words, either both acceptor & close_handler would be called once a connection successfully completes authentication, or error_handler would be called (with no call to acceptor or close_handler) when there's an error before or during authentication. Would this meet your requirement? Like the `acceptor()` call, the `SSHServerConnection` would be passed as an argument to the `close_handler()`.

**Comment by Kentzo** [comment]
> self._conn.create_task(do_some_cleanup(args))

I wasn't sure whether it was OK to use this method: `SSHConnection.__aexit__` doesn't await associated tasks so it's not similar to asyncio.TaskGroup.

> I guess one possible alternative would be to add another handler argument, alongside the acceptor and error_handler

Since it's the server that allocates and stores resources that need a cleanup, I'd expect this callback to be a bound method of the server as well.

**Comment by ronf** [comment]
While I understand your reasoning here, the AsyncSSH high-level API has moved away from applications needed to create custom subclasses. That made sense in the low-level callback-based world, but it's rarely necessary with the new AsyncSSH high-level API to need to subclass SSHServer and pass a custom server_factory argument. This might still be necessary if you are doing something which requires custom code for authentication, but otherwise it's generally possible to let AsyncSSH use the default SSHServer class and still be able to customize everything you need. Given that there's already a precedent here for handlers to be callables/coroutines, I don't really see adding a close handler that doesn't follow that same pattern.

It would be easy enough from you to call a bound method here. For instance, you could use lambda:

```python
    asyncssh.listen('', 8082, acceptor=lambda server: server.accept())
```

You could also declare a static method on your server subclass which takes a server instance as an argument and uses that to call other bound methods.

Regarding your point about not waiting for tasks to complete, AsyncSSH makes no guarantees about that right now. A call to wait_closed() will block until the SSH transport connection has closed, but there may still be tasks waiting on some method of the SSH connection or a session associated with it that could still be pending. These tasks should be unblocked by the close(), so typically they exit in fairly short order, but not necessarily before the wait_closed() returns.

Note that the cleanup function in general is scheduled as a "call_soon" call, and it's not asynchronous. So, any attempt to add a close handler would probably end up involving that being done via create_task(), as it's not possible to allow the current cleanup method to "block" and it would be quite a significant redesign to change that.

**Comment by Kentzo** [comment]
I think it's best to keep resources outside of AsyncSSH management then. I can find a workable solution with current implementation.
